Sources/SwiftLexicalLookup/LookupName.swift

```diff
@@ -0,0 +1,122 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+@_spi(Experimental) public enum LookupName {
+  /// Identifier associated with the name.
+  /// Could be an identifier of a variable, function or closure parameter and more
+  case identifier(IdentifiableSyntax, accessibleAfter: AbsolutePosition?)
+  /// Declaration associated with the name.
+  /// Could be class, struct, actor, protocol, function and more
+  case declaration(NamedDeclSyntax, accessibleAfter: AbsolutePosition?)
+
+  /// Syntax associated with this name.
+  @_spi(Experimental) public var syntax: SyntaxProtocol {
+    switch self {
+    case .identifier(let syntax, _):
+      syntax
+    case .declaration(let syntax, _):
+      syntax
+    }
+  }
+
+  /// Introduced name.
+  @_spi(Experimental) public var identifier: Identifier? {
+    switch self {
+    case .identifier(let syntax, _):
+      Identifier(syntax.identifier)
+    case .declaration(let syntax, _):
+      Identifier(syntax.name)
+    }
+  }
+
+  /// Point, after which the name is available in scope.
+  /// If set to `nil`, the name is available at any point in scope.
+  var accessibleAfter: AbsolutePosition? {
+    switch self {
+    case .identifier(_, let absolutePosition), .declaration(_, let absolutePosition):
+      absolutePosition
+    }
+  }
+
+  /// Checks if this name was introduced before the syntax used for lookup.
+  func isAccessible(at lookedUpSyntax: SyntaxProtocol) -> Bool {
+    guard let accessibleAfter else { return true }
+    return accessibleAfter <= lookedUpSyntax.position
+  }
+
+  /// Checks if this name refers to the looked up phrase.
+  func refersTo(_ lookedUpName: String) -> Bool {
+    guard let name = identifier?.name else { return false }
+    return name == lookedUpName
+  }
+
+  /// Extracts names introduced by the given `from` structure.
+  static func getNames(from syntax: SyntaxProtocol, accessibleAfter: AbsolutePosition? = nil) -> [LookupName] {
+    switch Syntax(syntax).as(SyntaxEnum.self) {
+    case .variableDecl(let variableDecl):
+      variableDecl.bindings.flatMap { binding in
+        getNames(from: binding.pattern, accessibleAfter: accessibleAfter)
+      }
```

Comment:
Thanks for finding this out! It doesn’t work indeed. As more special cases emerge, I think we could maybe introduce a bit more sophisticated way of extracting the names. We could move this logic to the syntax nodes themselves. This way, we would further unclutter scope implementations and unify internal name introduction to not be handled by the scopes. I'll try to go further with this idea and hopefully add a fix to #2748

Reply:
Ultimately, I don't think it will be actually necessary to change greatly the way we create names (at least in the next PR). Seems like making this example to work was a one line fix inside `LookupName.getNames` function. 

```swift
variableDecl.bindings.flatMap { binding in
  getNames(from: binding.pattern, accessibleAfter: accessibleAfter != nil ? binding.endPositionBeforeTrailingTrivia : nil)
}
```

I added this [test case](https://github.com/swiftlang/swift-syntax/blob/d5476066397a9a3c0fe7135cd04bd1983e42a28e/Tests/SwiftLexicalLookupTest/NameLookupTests.swift#L48-L70).